### PR TITLE
IMemoryScope for LG/Expression

### PR DESCRIPTION
**Pull request by boydc2014**
By allowing user to pass in a IMemoryScope, we can make the customization done in Expression libraray for AdaptiveDialog out of Expression library, by passing in a customized IMemoryScope. 

This is first step for #2430

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs [hunk 1]
```diff
@@ -902,170 +971,91 @@ private static (object value, string error) AccessIndex(object instance, int ind
...
-                modifiable = TryParseList(value, out var _);
-            }
-            else
-            {
-                // Modifiable object
-                modifiable = value is IDictionary<string, object>
-                    || value is IDictionary
-                    || value is JObject;
-                if (!modifiable)
+                catch (Exception e)
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs]: Why was this necessary?  We are trying to keep exceptions out of the expression evaluation--what is likely to throw an exception?

File: libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs [hunk 2]
```diff
@@ -902,170 +971,91 @@ private static (object value, string error) AccessIndex(object instance, int ind
...
                 {
-                    value = null;
+                    // the expression can't be fully merged as a path
+                    return (null, $"{expr.Children[0].ToString()} is not a valid path to set value");
                 }
+
+                var (value, error) = expr.Children[1].TryEvaluate(state);
+                return (state.SetValue(path, value), null);
+            }
+            catch (Exception e)
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs]: Same here--what is throwing an exception?

File: libraries/Microsoft.Bot.Builder.Expressions/SimpleObjectScope.cs [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Expressions
+{
+    internal class SimpleObjectScope : IMemoryScope
+    {
+        private object scope = null;
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Expressions/SimpleObjectScope.cs]: Please make sure you clean up all the visual studio (blue) warnings about style.  I suspect you are introducing some...

File: libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs (hunk 1 above)
Comment (boydc2014) [reply libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs]: the exception here introduced is because the IMemoryScope interface is like

object GetValue(string path) // here may through exception

if we make this interface return (result, error), it feels like a little bit like go-back to C-style, and actually this interface here is totally called by internal evaluation process, so i wrap the exception into (result, error) again here, so I can keep the interface simpler. 

Do you think that's worth it, or we can make the IMemoryScope interface GetValue return (result, error) also?

**Comment by chrimc62** [comment]
The issue is really how implementors handle part of a path not being present.  Because null is a valid value you cannot distinguish between the value being null and the path not being present.  Either we should:

  1.  Make clear to implementors that null can be a valid value or is what should be returned if part of a path is not present.  This loses the ability to distinguish between a valid null value and a path not present null value.
  2.  Throw an exception if a value is not present.  This doesn’t fly because for most rule evaluations we would be throwing exceptions.
  3.  Have some way of indicating the value is not present, i.e. return a value of null and an error of “path missing”.
In the old code we did 1.  I agree the exception backstop is reasonable there, but at a minimum, I think our implementations should not throw exceptions and we should add the information to the interface documentation.  I actually prefer doing 3 although it does make the interface a little more complex.  Otherwise for example, the UI has no way of saying “foobar” is not present in the memory vs. null is the value for “foobar”.  I think that is a better model.  Alternatively, maybe we should add a PathExists method.

**Comment by boydc2014** [comment]
> The issue is really how implementors handle part of a path not being present. Because null is a valid value you cannot distinguish between the value being null and the path not being present. Either we should: 1. Make clear to implementors that null can be a valid value or is what should be returned if part of a path is not present. This loses the ability to distinguish between a valid null value and a path not present null value. 2. Throw an exception if a value is not present. This doesn’t fly because for most rule evaluations we would be throwing exceptions. 3. Have some way of indicating the value is not present, i.e. return a value of null and an error of “path missing”. In the old code we did 1. I agree the exception backstop is reasonable there, but at a minimum, I think our implementations should not throw exceptions and we should add the information to the interface documentation. I actually prefer doing 3 although it does make the interface a little more complex. Otherwise for example, the UI has no way of saying “foobar” is not present in the memory vs. null is the value for “foobar”. I think that is a better model. Alternatively, maybe we should add a PathExists method. From: Dong Lei <notifications@github.com> Sent: Thursday, September 5, 2019 7:22 PM To: microsoft/botbuilder-dotnet <botbuilder-dotnet@noreply.github.com> Cc: Chris McConnell <chrimc@microsoft.com>; Review requested <review_requested@noreply.github.com> Subject: Re: [microsoft/botbuilder-dotnet] IMemoryScope for LG/Expression (#2468) @boydc2014 commented on this pull request.
> ________________________________ In libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fbotbuilder-dotnet%2Fpull%2F2468%23discussion_r321553843&data=02%7C01%7Cchrimc%40microsoft.com%7Cd017b4021ea9486965c808d732710195%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637033333253133840&sdata=ExB%2FtUoFGQc96iRbDWex%2BSB6coQ98NRobXGYyBDLoCE%3D&reserved=0>:
> - private static bool CanBeModified(object value, string property, int? expected)
> - { - var modifiable = false; - if (expected.HasValue) - { - // Modifiable list - modifiable = TryParseList(value, out var _); - } - else - { - // Modifiable object - modifiable = value is IDictionary<string, object> - || value is IDictionary - || value is JObject; - if (!modifiable) + catch (Exception e) the exception here introduced is because the IMemoryScope interface is like object GetValue(string path) // here may through exception if we make this interface return (result, error), it feels like a little bit like go-back to C-style, and actually this interface here is totally called by internal evaluation process, so i wrap the exception into (result, error) again here, so I can keep the interface simpler. Do you think that's worth it, or we can make the IMemoryScope interface GetValue return (result, error) also? — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fbotbuilder-dotnet%2Fpull%2F2468%3Femail_source%3Dnotifications%26email_token%3DAEGSCMM5ACNQN4CZ6E434SDQIG5EXA5CNFSM4ITBGEDKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCD3RFKA%23discussion_r321553843&data=02%7C01%7Cchrimc%40microsoft.com%7Cd017b4021ea9486965c808d732710195%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637033333253133840&sdata=MVS2NXvUzDbSLMvQQnH2hbnu1beAlYQsga11XLY2DTw%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAEGSCMJ6PEWLQYXGNSBJCATQIG5EXANCNFSM4ITBGEDA&data=02%7C01%7Cchrimc%40microsoft.com%7Cd017b4021ea9486965c808d732710195%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637033333253143838&sdata=CNcMQyvXus%2BG5nyOk%2FAa%2BCLzV%2Bwgfpvhqe61IZvqDbk%3D&reserved=0>.

I agree that the current interface are not clear about the semantic of a null return value. My initial thought was to take whatever returned as value, and let implementors throw exception for errors. I thought that would be more align to C# conventions, but not sure. 

I also agree that make the interface be explicit about value vs error would be more self document like this
```
(object value, string error) GetValue
```
i just from what you described, this approach is not against C# conventions, so it's good enough. 

In terms of "Path not exists error", the way you see it,  is it just a normal error that we can handle by the generic interface design above, or it's a specific type of error, that we want to special treatment?
I'm more thinking about the former, a generic approach, because my understanding of this design to delegate the path resolving to the implementors, who is supposed to handle path exists\not exists.
And the implementors can have it's own interpretation of the path. 

Given that, a "PathExists" methods sounds not necessary, because it makes the interface cares a little much about path exist problem, make our path resolving delegating not pure. So i can go ahead to make the interface looks like `(object value, string error)`, what do you think?

File: libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs (hunk 2 above)
Comment (boydc2014) [reply libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs]: i removed the exception handling, and let the interface explicit about value vs error.

**Comment by boydc2014** [comment]
> I quite like your solution to the sub-scope issue. The interface is nice and clean. Is there a separate PR for supporting the shortcuts and for setting any value?

Yes, next following PR is to move shorthands into dialog and create object in the middle.
